processor/src/decoder/aux_trace/block_hash_table.rs

```diff
@@ -27,177 +38,225 @@ impl<E: FieldElement<BaseField = Felt>> AuxColumnBuilder<E> for BlockHashTableCo
 
     /// Removes a row from the block hash table.
     fn get_requests_at(&self, main_trace: &MainTrace, alphas: &[E], i: usize) -> E {
-        let op_code_felt = main_trace.get_op_code(i);
-        let op_code = op_code_felt.as_int() as u8;
-
-        let op_code_felt_next = main_trace.get_op_code(i + 1);
-        let op_code_next = op_code_felt_next.as_int() as u8;
+        let op_code = main_trace.get_op_code(i).as_int() as u8;
 
         match op_code {
-            END => get_block_hash_table_removal_multiplicand(main_trace, i, alphas, op_code_next),
+            END => {
+                let op_code_next = main_trace.get_op_code(i + 1).as_int() as u8;
+
+                get_row_from_end(main_trace, i, alphas, op_code_next)
+            }
             _ => E::ONE,
         }
     }
 
     /// Adds a row to the block hash table.
     fn get_responses_at(&self, main_trace: &MainTrace, alphas: &[E], i: usize) -> E {
-        let op_code_felt = main_trace.get_op_code(i);
-        let op_code = op_code_felt.as_int() as u8;
+        let op_code = main_trace.get_op_code(i).as_int() as u8;
 
         match op_code {
-            JOIN => get_block_hash_table_inclusion_multiplicand_join(main_trace, i, alphas),
-            SPLIT => get_block_hash_table_inclusion_multiplicand_split(main_trace, i, alphas),
-            LOOP => get_block_hash_table_inclusion_multiplicand_loop(main_trace, i, alphas),
-            REPEAT => get_block_hash_table_inclusion_multiplicand_repeat(main_trace, i, alphas),
-            DYN => get_block_hash_table_inclusion_multiplicand_dyn(main_trace, i, alphas),
+            JOIN => {
+                let (left_child_row, right_child_row) = get_rows_from_join(main_trace, i, alphas);
+
+                // Note: this adds the 2 rows separately to the block hash table.
+                left_child_row * right_child_row
+            }
+            SPLIT => get_row_from_split(main_trace, i, alphas),
+            LOOP => get_row_from_loop(main_trace, i, alphas).unwrap_or(E::ONE),
+            REPEAT => get_row_from_repeat(main_trace, i, alphas),
+            DYN => get_row_from_dyn(main_trace, i, alphas),
             _ => E::ONE,
         }
     }
 }
 
+/// Builds a row for the block hash table. Since the block hash table is a virtual table, a "row" is
+/// a single field element representing all the columns (which is achieved by taking a random linear
+/// combination of all the columns). The columns are defined as follows:
+///
+/// - current_block_id: contains the ID of the current block. Note: the current block's ID is the
+///   parent block's ID from the perspective of the block being added to the table.
```

Comment:
nit: should we just use `parent_id` as in the docs? It was a bit easier for me to understand the logic of this virtual table using this name.

Reply:
Agreed, switched back to `parent_block_id`